Typecheck with mypy

dkfellows opened the pull request:
This is an adoption of #694 as a project-internal internal PR.

- [x] Types check
- [x] Unit tests pass
- [x] Integration tests pass

----

Tricky points:

- Various bits have to come from `typing_extensions` to work in 3.8; by 3.12 they're in Python itself.
- Pretty much every file requires `from __future__ import annotations` as the first line, which causes all annotations to be left as uninterpreted strings. (Without that, the syntax for unions would need to be much more wordy and things are bad enough as it is...)
- Type aliases usually need to be `Foo: TypeAlias = "Bar | Grill"` so that they get handled right.
- The new `type` declaration syntax of 3.12 isn't supported at all.
- Great care needs to be taken when doing specialization of a generic superclass in a subclass or using `cast` or `isinstance`; the type isn't auto-stringified there.
- Dataclasses require care:
  - Class variables must be annotated as such with `ClassVar`
  - You need to `@hydrate` the class before it is processed by `@dataclass` to make class variables be handled right.
  - The type of ordinary variables is sometimes stricter than the arguments that would be accepted. Handling the case where one narrows types (e.g., converting strings to enum members) during construction is *not* handled well by dataclasses; I ought to chase that up more generally...
- I've *started* to tame the wild dictionaries with `TypedDict`s, but that job isn't complete and I've no idea how much effort should be put into it; this has been one of the slowest parts: identifying where types could be narrowed.
- Dictionaries that come from JSON should more likely be `dict[str, Any]` than straight `Dict`; JSON doesn't allow non-string keys.
- Imports guarded by `typing.TYPE_CHECKING` have no constraints based on name resolution order... but can only be used in uninterpreted types (as the code isn't actually executed by Python).

dkfellows committed:
Add workflow for running mypy as an on-push workflow

dkfellows committed:
Type improvements

dkfellows committed:
Gradually...

dkfellows committed:
ObjectList is mostly right now

dkfellows committed:
That's element.py now type-checking

dkfellows committed:
types for loop.py

dkfellows committed:
command_files.py now typechecks

dkfellows committed:
commands.py now typechecks

dkfellows committed:
Working towards typing of actions.py

dkfellows committed:
More typing for actions.py

dkfellows committed:
Types for environment.py

dkfellows committed:
Finally, got parameters.py to typecheck...

dkfellows committed:
Utils done

dkfellows committed:
forgot a space

dkfellows committed:
Most of the way through tasks

dkfellows committed:
task.py now typechecks!

dkfellows committed:
And now I'm fairly sure it's about right

dkfellows committed:
Fix task_schema.py

dkfellows committed:
Fix test_utils

dkfellows committed:
as_object overloads

dkfellows committed:
Starting to really reduce errors in workflow.py

dkfellows committed:
Many more errors fixed in workflow.py

dkfellows committed:
Added missing abstract methods

dkfellows committed:
No more errors in workflow.py

dkfellows committed:
Core now all typechecks!

dkfellows committed:
Starting to fix problems in submission subsystem

dkfellows committed:
Towards all submission code

dkfellows committed:
both core and scheduler now typecheck fully

dkfellows committed:
Starting to clean up the persistence types

dkfellows committed:
Types for persistence

dkfellows committed:
Types for config

dkfellows committed:
Various small bits

dkfellows committed:
Roll up a bunch more changes to make types more consistent

dkfellows committed:
We're making progress!

dkfellows committed:
Whole sdk typechecks!

dkfellows committed:
Fixes so everything passes

dkfellows committed:
Workaround for mypy alternating between understanding ruamel and not
See python/mypy#7276 and https://sourceforge.net/p/ruamel-yaml/tickets/328/ for a whole round of finger-pointing (or https://stackoverflow.com/a/78386020/301832 for a summary)

dkfellows committed:
Reducing the number of FIXMEs

dkfellows committed:
more typed dicts

dkfellows committed:
More tweaks

dkfellows committed:
Use protocols correctly

dkfellows committed:
Cleaning up

Reducing the number of casts, fixing types, defining complex enums in the officially documented way, etc.

dkfellows committed:
Final BaseApp type thunks

dkfellows committed:
Optimise dynamic sublist creation

dkfellows committed:
A bunch more improvements

I tried temporarily turning on checking with pyright and turning on --disallow-untyped-defs for mypy, but the noise was considerable.

dkfellows committed:
Gradually trying to get definite types for all properties

dkfellows committed:
Useful types for config

dkfellows committed:
Typing the CLI

dkfellows committed:
Cleaning up some more types

dkfellows committed:
Reducing the number of raw untyped dicts

dkfellows committed:
The tests are another source of type information

dkfellows committed:
More type improvements

dkfellows committed:
Typing the unit tests

dkfellows committed:
Integration tests, misc

dkfellows committed:
update poetry.lock

dkfellows committed:
Add type stubs

dkfellows committed:
Tweak

dkfellows committed:
Read flags in a way that doesn't trigger a typecheck

dkfellows committed:
Fix stringified annotations and dataclasses

https://stackoverflow.com/q/73091205/301832 hints at what is going on (and it's really due a bunch of people in Python pointing their fingers at each other due to a fight over the semantics of annotations) but the fix I've done is a bit more conservative and works better because it doesn't try to do a full resolution of everything. It's all awful.

dkfellows committed:
Fixing some tests

dkfellows committed:
Fix more problems

dkfellows committed:
Redo WorkflowTask._get_merged_parameter_data to not trigger a  complex test failure

dkfellows committed:
Many more tests pass

dkfellows committed:
doc

dkfellows committed:
More test fixes

dkfellows committed:
Remove a few debugging prints!

dkfellows committed:
Well that was a dumb bug on my part...

dkfellows committed:
Fix more tests

dkfellows committed:
One more silly error

dkfellows committed:
Fix some equality tests

dkfellows committed:
More fixes

dkfellows committed:
A bunch more fixes, including making something into a TypedDict

dkfellows committed:
Well, that was a silly error!

dkfellows committed:
Fix my path normalization errors

dkfellows committed:
I don't understand nesting orders completely yet

dkfellows committed:
Cleaning up some more types

dkfellows committed:
Unit tests work! Added doc note about a critical cast that ends up causing dot products instead of cross products.

dkfellows committed:
Reduce number of FIXMEs and raw Dict types

dkfellows committed:
Black really didn't like embedded double quoted strings in double quoted strings

dkfellows committed:
black

dkfellows committed:
Another black-touched file

dkfellows committed:
Merge branch 'develop' into typecheck-with-mypy

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Missed a spot

dkfellows committed:
Factor out class which moved between Python versions

There will probably be more...

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Tweak

dkfellows committed:
black

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Gradually fixing issues

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Aliases need a little more care

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Use typing_extensions directly

It handles the "typing already does this" for us.

dkfellows committed:
More fixes

dkfellows committed:
I missed some places in an earlier commit

dkfellows committed:
More

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Parameters to specialize a subclass of a generic class need to be carefully done; they're not auto-stringed annotations.

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
NoneType is a 3.10-ism

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
More care with cast

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
utcnow is deprecated

dkfellows committed:
Fix bug in integration test with HDF5 saving

ParameterValue classes were getting lost. No idea why!

dkfellows commented:
I've been investigating a bug, when handling the EAR end step workflow callback (i.e., not in the main script itself), parameters would not have their value class set when processing the results. I'm not at all sure why this would be so (as the schema had been parsed and the parameter class modules loaded) but it was resulting in the reading out of values from an HDF5 file not working. I've resolved it by doing the forcing at that point (because the alternative is a definite failure! New method `Parameter._force_value_class()`) but it's really odd because I can't see why the parameters are wrong in the first place.

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Tweaking

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Chiselling away some more of the raw lists and dicts

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Tidy up usage of valida a bit

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Clean up

These changes are due to using flake8

dkfellows committed:
Two more spots that don't need f-strings

aplowman commented on hpcflow/sdk/app.py:
```diff
@@ -780,43 +1289,39 @@ def _load_scripts(self):
-            if i.suffix in (".yaml", ".yml", ".json", ".jsonc"):
-                templates[i.stem] = i
+        for i in resources.files(pkg).iterdir():
+            p = Path(i.name)
```
This should be `p = Path(i)` I think. Otherwise, `templates` is just the file names.

aplowman commented on hpcflow/sdk/core/actions.py:
```diff
@@ -1115,26 +1296,26 @@ class Action(JSONLike):
-    _script_data_formats = ("direct", "json", "hdf5")
+    _script_data_formats: ClassVar[tuple[str, ...]] = ("direct", "json", "hdf5")
 
     def __init__(
```
I'm not sure where the problem is but there seems to be an issue with "expanded" actions. For example: 

```py
import hpcflow.app as hf

hf.task_schemas.dummy_task_1.actions[0].script_data_in # empty dict; fine
hf.task_schemas.dummy_task_1.actions[1].script_data_in # --> AttributeError: 'Action' object has no attribute 'script_data_in'

```

aplowman commented on hpcflow/sdk/typing.py:
```diff
@@ -1,9 +1,62 @@
-E_idx_type = Tuple[int, int]
-EI_idx_type = Tuple[int, int, int]
-EAR_idx_type = Tuple[int, int, int, int, int]
+E_idx_type: TypeAlias = "tuple[int, int]"
```
I think these three (`E_idx_type`, `EI_idx_type`, `EAR_idx_type`) can be removed, since they are not used anywhere as far as I can tell.

dkfellows replied:
Seems so. The exact tuple corresponding to `E_idx_type` was used internally in a few places, but not enough to justify a global alias, and the others simply weren't used at all (with the same semantics of indices)

aplowman commented on hpcflow/sdk/typing.py:
```diff
@@ -1,9 +1,62 @@
+DataIndex: TypeAlias = "dict[str, int | list[int]]"
+"""
+The type of indices to data. These are *normally* dictionaries of integers,
+but can have leaves being lists of integers when dealing with iterations.
```
The data index values are lists of integers not for iterations but for element groups (i.e. when a downstream element uses outputs from multiple upstream elements, rather than just a single upstream element).

dkfellows replied:
Thanks for clearing up my misunderstanding

aplowman committed:
fix: replace call to non-existent method

dkfellows committed:
Neater App awareness

I've been thinking about it, and factoring out the awareness of the BaseApp instance into its own simple class. Most classes then pick that up through inheritance.

This also makes references to the BaseApp be mainly consistently by _app; they're an implementation detail.

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Why did this work before?

dkfellows commented:
Do #699 first; it will be easier that way round.

dkfellows committed:
Merge branch 'develop' into typecheck-with-mypy

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Fix type errors

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
No, Black, you got that wrong; stop sabotaging me

dkfellows committed:
Schema says parameters don't have names

Properly, their names are the last element of the path that leads to them, but that's not known to the Parameter class itself.

dkfellows committed:
Not quite sure what was not working

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Add missing docs

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
More missing docs

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
More docs

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Improve documentation of BaseApp

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Cleaning up

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Document return values of major ops

dkfellows commented:
This looks pretty good to me now. I'd love the documentation of the typed dictionaries to be better... but I'm not fixing Python itself so it will just have to be what it is.

dkfellows committed:
Add missing docs, move some to better locations

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Tighten up resource access, type more args

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Reducing the number of untyped arguments

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Add more missing types, group TypedDicts in each file

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Add note about type

dkfellows committed:
Now I know why TypedDict comes from typing_extensions

dkfellows committed:
Import in right place

dkfellows committed:
Move the typed dicts into files that just handle types

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows commented:
These are bizarre failures, FWIW. I can only reproduce them locally by not having Python on the path.

dkfellows committed:
Something is really mis-handling type parameters

Make them less obnoxious in their docs.
Add precompilation of REs where possible. Note that that isn't always possible...

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Try to make things a little cleaner

Still can't see why 3.8 suddenly started failing; the test logs are all "computer says no" level of informative.

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Tinkering

Pull most enums into separate files for simpler internal dependency graphs. Some more efficient ways of handling iterations over lists of things.

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Not finding a partition is OK

dkfellows committed:
Compat

dkfellows committed:
Bake most exception messages into the exceptions

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Oops, too tight a type

dkfellows committed:
Adding types to more untyped arguments

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Some efficiency improvements

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Optimisations to reduce the number of copies

Also fix a casting blunder (by me!) in actions that was making integration tests fail.

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Reducing the twistedness of the code

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Further trying to reduce both the amount of work and the visibility scope of methods

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Tighening up the code, reducing number of confusing single-letter variables

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Reducing use of single-letter variables

Expecially i, which has previously been used with ever so many meanings...

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Final round of cleanup

Note the switch to TypeIs, which has better semantics than TypeGuard

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Fix silly error

dkfellows committed:
Handle sets of IDs where possible

Avoids quite a bit of conversion and sorting

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
More immutable types...

Including an immutable hashable mapping for loop indices. Also some better ways to handle splitting and replacements and trimming.

dkfellows committed:
Merge branch 'develop' into typecheck-with-mypy

hpcflow-actions committed:
pre-commit fixes [skip ci]

aplowman committed:
test: add failing tests for `Task.is_input_type_required` when an input file generator is used

aplowman commented on hpcflow/sdk/core/actions.py:
```diff
@@ -2377,21 +2458,21 @@ def get_precise_scope(self) -> app.ActionScope:
 
     def is_input_type_required(
-        self, typ: str, provided_files: List[app.FileSpec]
+        self, typ: str, provided_files: Container[FileSpec]
```
I think using a set instead of a list for `provided_files` means the membership test `ifg.input_file not in provided_files` below fails with `TypeError: unhashable type: 'FileSpec'`. I have just committed two two failing tests (8fb13758d44826f2bc595bada746b57c95a19dd2) that test for the correct behaviour.

dkfellows replied:
Technically, the requirement here is that `provided_files` supports `in`, which is a `Container` (one of the abstract supertypes of both lists and sets) and your change actually makes the problem be at:

https://github.com/hpcflow/hpcflow-new/blob/3c68e1891e80cc97450cff67387715b75793d000/hpcflow/sdk/core/task.py#L1443

I'll fix this by making `FileSpec` be a hashable type (by `label` and `name`, the fields used in the equality test; also similarly for `FileNameSpec`); there's no real reason not to.

dkfellows replied:
Which turned out to be _slightly_ tricky to do. Fortunately, it seems that `FileNameSpec` is treated as a read-only class so it doesn't need a `list` inside (a `tuple` does fine). I couldn't find any examples of `FileNameSpec.args` being non-empty, but I'll assume it's possible.

dkfellows committed:
Merge branch 'develop' into typecheck-with-mypy

dkfellows committed:
Remove unused types

dkfellows committed:
Data index meaning corrected

dkfellows committed:
Make FileSpec hashable

dkfellows committed:
Filter internal variables from serialisations

Introduce a common abstract model for a renderable filename fragment.

dkfellows committed:
Merge branch 'develop' into typecheck-with-mypy

hpcflow-actions committed:
pre-commit fixes [skip ci]

dkfellows committed:
Add missing type stubs

dkfellows committed:
Subtle behaviour change between Python versions

dkfellows committed:
Merge branch 'develop' into typecheck-with-mypy

dkfellows committed:
Merge branch 'develop' into typecheck-with-mypy
